### PR TITLE
[MC] Eliminate two symbol-related hash maps

### PR DESCRIPTION
**Pull request by aengelke**
Previously, a symbol insertion requires (at least) three hash table operations:

- Lookup/create entry in Symbols (main symbol table)
- Lookup NextUniqueID to deduplicate identical temporary labels
- Add entry to UsedNames, which is also used to serve as storage for the symbol name in the MCSymbol.

All three lookups are done with the same name, so combining these into a single table reduces the number of lookups to one. Thus, a pointer to a symbol table entry can be passed to createSymbol to avoid a duplicate lookup of the same name.

The new symbol table entry value is placed in a separate header to avoid including MCContext in MCSymbol or vice versa.

http://llvm-compile-time-tracker.com/compare.php?from=846e47e7b880bcf6b8f5773fe0fe236d486f3239&to=a1e780d1eea6be4d79d40d825d2b3e081e765c55&stat=instructions:u

### CODE REVIEW

File: llvm/include/llvm/MC/MCSymbolTableEntry.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  /// Whether the name associated with this value is used for a symbol. This is
+  /// not necessarily true: sometimes, we use a symbol table value without an
+  /// associated symbol for accessing NextUniqueID when a suffix is added to a
+  /// name. However, Used might be true even if Symbol is nullptr: temporary
+  /// named symbols are not added to the symbol table.
+  bool Used = false;
+};
+
+/// MCContext stores MCSymbolTableValue in a string map. To avoid redundant
```
Comment (MaskRay) [inline llvm/include/llvm/MC/MCSymbolTableEntry.h:38]: Should the comment be moved to MCSymbol.h? Or leave here just a comment like "See MCSymbol::operator new".

**Review by MaskRay — APPROVED** [review]
Very nice! I stared at the code before but did not realize it is a bottleneck and should be fixed.

File: llvm/lib/MC/MCContext.cpp [hunk 2]
```diff
@@ -228,7 +226,11 @@ MCSymbol *MCContext::getOrCreateLSDASymbol(const Twine &FuncName) {
                            FuncName);
 }
 
-MCSymbol *MCContext::createSymbolImpl(const StringMapEntry<bool> *Name,
+MCSymbolTableEntry &MCContext::getSymbolTableEntry(StringRef Name) {
+  return *Symbols.insert(std::make_pair(Name, MCSymbolTableValue{})).first;
```
Comment (MaskRay) [inline llvm/lib/MC/MCContext.cpp:230]: Perhaps `try_emplace(Name, MCSymbolTableValue{})`

**Comment by MaskRay** [comment]
How about making the following change then removing the `Entry` parameter from `createSymbol`?
```
diff --git i/llvm/lib/MC/MCContext.cpp w/llvm/lib/MC/MCContext.cpp
index 7ac44f41186b..535e6d0313d7 100644
--- i/llvm/lib/MC/MCContext.cpp
+++ w/llvm/lib/MC/MCContext.cpp
@@ -206,4 +206,9 @@ MCSymbol *MCContext::getOrCreateSymbol(const Twine &Name) {
   MCSymbolTableEntry &Entry = getSymbolTableEntry(NameRef);
-  if (!Entry.second.Symbol)
-    Entry.second.Symbol = createSymbol(&Entry, NameRef, false, false);
+  if (!Entry.second.Symbol) {
+    bool IsTemporary = false;
+    if (!SaveTempLabels)
+      IsTemporary = NameRef.starts_with(MAI->getPrivateGlobalPrefix());
+    Entry.second.Used = true;
+    Entry.second.Symbol = createSymbolImpl(&Entry, IsTemporary);
+  }

@@ -229,3 +234,3 @@ MCSymbol *MCContext::getOrCreateLSDASymbol(const Twine &FuncName) {
 MCSymbolTableEntry &MCContext::getSymbolTableEntry(StringRef Name) {
-  return *Symbols.insert(std::make_pair(Name, MCSymbolTableValue{})).first;
+  return *Symbols.try_emplace(Name, MCSymbolTableValue{}).first;
 }
```

**Comment by aengelke** [comment]
> How about making the following change then removing the Entry parameter from createSymbol?

This doesn't work that easily, as private user-visible symbols do get renamed. I refactored the code a bit so that the method is now named `createRenamableSymbol`. For user-visible symbols that occur more than once (unlikely to happen), we now do two lookups, but I think that's tolerable.

**Pull request merged**